serviceregistry-server/src/main/java/no/difi/meldingsutveksling/serviceregistry/auth/TokenValidator.java

```diff
@@ -0,0 +1,114 @@
+import com.nimbusds.jwt.proc.ConfigurableJWTProcessor;
+import com.nimbusds.jwt.proc.DefaultJWTProcessor;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import no.difi.meldingsutveksling.serviceregistry.config.ServiceregistryProperties;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.net.URL;
+import java.text.ParseException;
+
+@Slf4j
+@Component
+public class TokenValidator {
+
+    /**
+     * The signing algorithm.
+     */
+    private static final JWSAlgorithm SIGNING_ALGORITHM = JWSAlgorithm.RS256;
+
+    /**
+     * The expected token issuer (from configuration)
+     */
+    private final String tokenIssuer;
+
+    /**
+     * The JWKS URI (from configuration)
+     */
+    private final String jwksUri;
+
+    public TokenValidator(ServiceregistryProperties properties) {
+        ServiceregistryProperties.Auth authorizationProperties = properties.getAuth();
+        this.tokenIssuer = authorizationProperties.getIssuer();
+        this.jwksUri = authorizationProperties.getJwksUri();
+    }
+
+    public Result validate(String token) throws IOException {
+        JWSObject jwsToken;
+        try {
+            jwsToken = JWSObject.parse(token);
+        } catch (ParseException e) {
+            log.error("Invalid authorization token.");
+            return new Result(false, null);
+        }
+        String algorithm = jwsToken.getHeader().getAlgorithm().getName().toUpperCase();
+        if (!validateSigningAlgorithm(algorithm)) {
+            return new Result(false, null);
+        }
+        ConfigurableJWTProcessor jwtProcessor = new DefaultJWTProcessor();
+        JWKSource keySource = new RemoteJWKSet(new URL(jwksUri));
+        JWSKeySelector keySelector = new JWSVerificationKeySelector(SIGNING_ALGORITHM, keySource);
+        jwtProcessor.setJWSKeySelector(keySelector);
+        SecurityContext context = null;
+        try {
+            JWTClaimsSet claims = jwtProcessor.process(token, context);
+            return new Result(validateTokenIssuer(claims), claims);
+        } catch (ParseException e) {
+            log.error("Unable to parse the authorization token", e);
+        } catch (BadJOSEException e) {
+            log.error("Invalid authorization token signature", e);
+        } catch (JOSEException e) {
+            log.error("Unable to process the authorization token signature", e);
+        }
+        return new Result(false, null);
+    }
+
+    @Value
+    public class Result {
+        boolean valid;
+        JWTClaimsSet claims;
+    }
+
+    /**
+     * Check whether the token has been released by the expected issuer
+     */
+    private Boolean validateTokenIssuer(JWTClaimsSet claims) {
+        String issuer = claims.getIssuer();
+        if (issuer == null) {
+            log.error("The authorization token doesn't have an issuer (iss)");
+            return false;
+        }
+        if (issuer.toLowerCase().equals(this.tokenIssuer)) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Replace these toUpperCase()/toLowerCase() and equals() calls with a single equalsIgnoreCase() call. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar-app01.difi.local:9000/coding_rules#rule_key=squid%3AS1157)